File: GPM/workbench/web/src/api/endpoints/size.ts
import type { WorkbenchResponse } from "../gpm-api";
import { request } from "../request";

export const sizeEndpoints = {
  sizeEncodeWord: (
    contentHash: string,
    profile: string,
    fallback?: { original: string; normalized: string; substance: number; perm_index: number },
  ) =>
    request<WorkbenchResponse>("/api/size/encode-word", {
      method: "POST",
      body: JSON.stringify({
        content_hash: contentHash,
        profile,
        fallback: fallback
          ? {
              original: fallback.original,
              normalized: fallback.normalized,
              substance: fallback.substance,
              perm_index: fallback.perm_index,
            }
          : null,
      }),
    }),

  sizeEncodeBatch: (wordHashes: string[], profile: string) =>
    request<WorkbenchResponse>("/api/size/encode-batch", {
      method: "POST",
      body: JSON.stringify({ word_hashes: wordHashes, profile }),
    }),

  sizeDecode: (
    contentHash: string,
    profile: string,
    fallback?: { substance: number; perm_index: number },
  ) =>
    request<WorkbenchResponse>("/api/size/decode", {
      method: "POST",
      body: JSON.stringify({
        content_hash: contentHash,
        profile,
        fallback: fallback
          ? { substance: fallback.substance, perm_index: fallback.perm_index }
          : null,
      }),
    }),
};

File: GPM/workbench/api/routes/size.py
"""Size compare API routes."""

from __future__ import annotations

from fastapi import APIRouter

from api.schemas.common import WorkbenchResponse
from api.schemas.requests import (
    SizeDecodeRequest,
    SizeEncodeBatchRequest,
    SizeEncodeWordRequest,
)
from api.size_service import (
    compare_decode_cached,
    compare_encode_batch_cached,
    compare_encode_word_cached,
)

router = APIRouter(prefix="/api/size", tags=["size"])


@router.post("/encode-word", response_model=WorkbenchResponse)
def size_encode_word(req: SizeEncodeWordRequest) -> WorkbenchResponse:
    fallback = req.fallback.model_dump() if req.fallback is not None else None
    result = compare_encode_word_cached(
        content_hash=req.content_hash,
        profile=req.profile,
        fallback=fallback,
    )
    return WorkbenchResponse(result=result)


@router.post("/encode-batch", response_model=WorkbenchResponse)
def size_encode_batch(req: SizeEncodeBatchRequest) -> WorkbenchResponse:
    result = compare_encode_batch_cached(word_hashes=req.word_hashes)
    return WorkbenchResponse(result=result)


@router.post("/decode", response_model=WorkbenchResponse)
def size_decode(req: SizeDecodeRequest) -> WorkbenchResponse:
    fallback = req.fallback.model_dump() if req.fallback is not None else None
    result = compare_decode_cached(
        content_hash=req.content_hash,
        profile=req.profile,
        fallback=fallback,
    )
    return WorkbenchResponse(result=result)
